src/renderers/shared/utils/ReactErrorUtils.js

```diff
@@ -155,84 +173,93 @@ if (__DEV__) {
+      // the error event at all.
       let didError = true;
+
+      // Create an event handler for our fake event. We will synchronously
+      // dispatch our fake event using `dispatchEvent`. Inside the handler, we
+      // call the user-provided callback.
       const funcArgs = Array.prototype.slice.call(arguments, 3);
-      const boundFunc = function() {
+      const onFakeEvent = function() {
         func.apply(context, funcArgs);
         didError = false;
       };
-      const onFakeEventError = function(event) {
+
+      // Create a global error event handler. We use this to capture the value
+      // that was thrown. It's possible that this error handler will fire more
+      // than once; for example, if non-React code also calls `dispatchEvent`
+      // and a handler for that event throws. We should be resilient to most of
+      // those cases. Even if our error event handler fires more than once, the
+      // last error event is always used. If the callback actually does error,
+      // we know that the last error event is the correct one, because it's not
+      // possible for anything else to have happened in between our callback
+      // erroring and the code that follows the `dispatchEvent` call below. If
+      // the callback doesn't error, but the error event was fired, we know to
+      // ignore it because `didError` will be false, as described above.
+      let error;
+      // Use this to track whether the error event is ever called.
+      let didSetError = false;
+
+      const onError = function(event) {
         error = event.error;
-        if (preventDefault) {
-          event.preventDefault();
-        }
+        didSetError = true;
       };
 
+      // Create a fake event type. We add `depth` to the event name so that
+      // nested `invokeGuardedCallback` calls do not clash. Otherwise, a nested
+      // call would trigger the fake event handlers of any call higher in
+      // the stack.
+      depth++;
       const evtType = `react-${name ? name : 'invokeguardedcallback'}-${depth}`;
-      window.addEventListener('error', onFakeEventError);
-      fakeNode.addEventListener(evtType, boundFunc, false);
-      const evt = document.createEvent('Event');
 
+      // Attach our event handlers
+      window.addEventListener('error', onError);
+      fakeNode.addEventListener(evtType, onFakeEvent, false);
+
+      // Syncrhonously dispatch our fake event. If the user-provided function
+      // errors, it will trigger our global error handler.
+      const evt = document.createEvent('Event');
       evt.initEvent(evtType, false, false);
       fakeNode.dispatchEvent(evt);
+
       if (didError) {
+        if (!didSetError) {
+          // The callback errored, but the error event never fired.
+          error = new Error(
+            'An error was thrown inside one of your components, but React ' +
+              "doesn't know what it was. This is likely due to browser " +
```

Comment:
"Clever" is a bit self congratulatory and probably doesn't work for the case where we can't provide useful info